### PR TITLE
Inference tutorial - Part 3 of e2e series [WIP]

### PR DESCRIPTION
**Pull request by jainapurva**

### CODE REVIEW

File: docs/source/inference.rst [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+
+.. code-block:: bash
+
+    # baseline
+    python benchmarks/benchmark_latency.py --input-len 256 --output-len 256 --model microsoft/Phi-4-mini-instruct --batch-size 1
+
+    # float8dq
+    VLLM_DISABLE_COMPILE_CACHE=1 python benchmarks/benchmark_latency.py --input-len 256 --output-len 256 --model pytorch/Phi-4-mini-instruct-float8dq --batch-size 1
+
+**Serving Benchmarking**:
```
Comment (jerryzh168) [inline docs/source/inference.rst]: both of these can be categorized into **Performance Benchmarking** and we can have a separate section for **Memory Benchmarking** I think

File: docs/source/inference.rst [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
++----------------------------+----------------+------------------------------+
+| latency (batch_size=128)   | 3.1s           | 2.72s (1.14x speedup)        |
++----------------------------+----------------+------------------------------+
+| serving (num_prompts=1)    | 1.35 req/s     | 1.57 req/s (1.16x speedup)   |
++----------------------------+----------------+------------------------------+
+| serving (num_prompts=1000) | 66.68 req/s    | 80.53 req/s (1.21x speedup)  |
++----------------------------+----------------+------------------------------+
+
+
+Sparsity Integration
```
Comment (jerryzh168) [inline docs/source/inference.rst]: this should not be a separate section I think, it can be merged into `Float8 Dynamic Quantization` section, and just mention for more quantization/sparsity, please see https://huggingface.co/docs/transformers/main/en/quantization/torchao

File: docs/source/inference.rst [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+        print("\nGenerated Outputs:\n" + "-" * 60)
+        for output in outputs:
+            prompt = output.prompt
+            generated_text = output.outputs[0].text
+            print(f"Prompt:    {prompt!r}")
+            print(f"Output:    {generated_text!r}")
+            print("-" * 60)
+
+
+Serving Quantized Models
```
Comment (jerryzh168) [inline docs/source/inference.rst]: this section can come before `Inference with vLLM` since it's simpler I think

File: docs/source/inference.rst [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+- **Accuracy**: Maintained within 5-10% of original model on most benchmarks
+
+
+Conclusion
+##########
+
+This tutorial demonstrated how torchao's quantization and sparsity techniques integrate seamlessly across the entire ML deployment stack:
+
+- **HuggingFace Transformers** provides easy model loading with torchao quantization
+- **vLLM** leverages torchao's optimized kernels for high-throughput serving
```
Comment (jerryzh168) [inline docs/source/inference.rst]: also add lm-eval for model quality evaluation

File: docs/source/inference.rst [hunk 5]
```diff
@@ -0,0 +1,463 @@
+##################################################
+Inference Tutorial: From Quantization to Serving
+##################################################
+
+This tutorial demonstrates how to perform post-training quantization and deploy models for inference:
+
+1. :ref:`Post-training Quantization with HuggingFace`: Using float8 dynamic quantization with HuggingFace integration
+2. :ref:`Sparsity Integration`: Combining sparsity with quantization for additional speedups
+3. :ref:`High-throughput Serving with vLLM`: Deploying quantized models with vLLM
+4. :ref:`Mobile Deployment with Executorch`: Lowering to ExecuTorch for on-device inference
```
Comment (jerryzh168) [inline docs/source/inference.rst]: we can add the model quality eval and performance benchmarking here as well

Comment (jainapurva) [reply docs/source/inference.rst]: Updated structure

File: docs/source/inference.rst [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        return_tensors="pt",
+    ).to("cuda")
+    generated_ids = quantized_model.generate(**inputs, max_new_tokens=128)
+    output_text = tokenizer.batch_decode(
+        generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
+    )
+    print("Response:", output_text[0][len(prompt):])
+
+
+[Optional] Float8 Dynamic Quantization + Semi-structured (2:4) sparsity
```
Comment (jainapurva) [inline docs/source/inference.rst]: @jerryzh168 @jcaip Does this look good? Should I keep sparsity as a optional section or just mention it in note

Comment (jerryzh168) [reply docs/source/inference.rst]: can we just add to huggingface torchao page?

File: docs/source/inference.rst [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+    # Evaluate baseline model
+    lm_eval --model hf --model_args pretrained=microsoft/Phi-4-mini-instruct --tasks hellaswag --device cuda:0 --batch_size 8
+
+    # Evaluate torchao-quantized model (float8dq)
+    lm_eval --model hf --model_args pretrained=pytorch/Phi-4-mini-instruct-float8dq --tasks hellaswag --device cuda:0 --batch_size 8
+
+Memory Benchmarking
+--------------------
+
+**Memory Usage Comparison**:
```
Comment (jerryzh168) [inline docs/source/inference.rst]: nit: remove

File: docs/source/inference.rst [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    pip install torchao
+
+Serving Quantized Models
+========================
+
+.. code-block:: bash
+
+    vllm serve pytorch/Phi-4-mini-instruct-float8dq --tokenizer microsoft/Phi-4-mini-instruct -O3
+
+Inference with vLLM
```
Comment (jerryzh168) [inline docs/source/inference.rst]: should we move this after `Inference with Transformers`

File: docs/source/inference.rst [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+
+Serving
+#######
+
+High-throughput Serving with vLLM
+---------------------------------
+
+vLLM automatically leverages torchao's optimized kernels when serving quantized models, providing significant throughput improvements.
+
+Setting up vLLM with Quantized Models
```
Comment (jerryzh168) [inline docs/source/inference.rst]: nit: this doesn't have to be a new section I think

File: docs/source/inference.rst [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+            prompt = output.prompt
+            generated_text = output.outputs[0].text
+            print(f"Prompt:    {prompt!r}")
+            print(f"Output:    {generated_text!r}")
+            print("-" * 60)
+
+Performance Breakdown
+=====================
+
+When using vLLM with torchao:
```
Comment (jerryzh168) [inline docs/source/inference.rst]: this is not a comprehensive list, probably just remove, do we have a exhaustive list of all the techniques that we support?

**Comment by andrewor14** [comment]
Hi @jainapurva, by the way I'm adding a `serving.rst` here: https://github.com/pytorch/ao/pull/2394. It uses the same template as parts 1 and 2. After that's landed, do you mind updating your PR to use that file instead? Right now it's a blank page with the template:

Screenshot 2025-06-17 at 5 48 14 PM

File: docs/source/inference.rst (hunk 10 above)
Comment (jainapurva) [reply docs/source/inference.rst]: I don't think we've a comprehensive list. If we decide to make it, that could be another doc page or readme

File: docs/source/inference.rst [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+
+    from torchao.quantization import Float8DynamicActivationFloat8SemiSparseWeightConfig
+
+    # Combine sparsity with int4 quantization - both optimized by torchao
+    quant_config = Float8DynamicActivationFloat8SemiSparseWeightConfig()
+    quantization_config = TorchAoConfig(quant_type=quant_config)
+
+    # Load a pre-sparsified checkpoint
+    model = AutoModelForCausalLM.from_pretrained(
+        "nm-testing/Meta-Llama-3.1-8B-Instruct-W4A16-G128-2of4",  # 2:4 sparse model
```
Comment (jcaip) [inline docs/source/inference.rst]: We should use this checkpoint instead: https://huggingface.co/RedHatAI/Sparse-Llama-3.1-8B-2of4 for FP8

File: docs/source/inference.rst (hunk 8 above)
Comment (jcaip) [reply docs/source/inference.rst]: cc @jainapurva I think if vLLM is our recommended serving solution, this should go before transformers.

File: docs/source/inference.rst (hunk 10 above)
Comment (jerryzh168) [reply docs/source/inference.rst]: I think we can remove this section in that case

File: docs/source/inference.rst (hunk 6 above)
Comment (jainapurva) [reply docs/source/inference.rst]: Remove it from here, and just add keep the note for hf-torchao page ?

File: docs/source/inference.rst [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+
+Post-training Quantization with HuggingFace
+############################################
+
+HuggingFace Transformers provides seamless integration with torchao quantization. The ``TorchAoConfig`` automatically applies torchao's optimized quantization algorithms during model loading.
+
+Float8 Dynamic Quantization
+------------------------------
+
+Float8 dynamic quantization shows 36% reduction in model size with minimal accuracy loss:
```
Comment (jerryzh168) [inline docs/source/inference.rst]: nit: we can remove this I think, it's specific to one model, probably don't need to mention here

File: docs/source/inference.rst (hunk 6 above)
Comment (jerryzh168) [reply docs/source/inference.rst]: Yeah

File: docs/source/inference.rst [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+        do_sample=False
+    )
+
+    response = tokenizer.decode(outputs[0][inputs.shape[1]:], skip_special_tokens=True)
+    print(response)
+
+.. note::
+    For more information on supported quantization and sparsity configurations, see `HF-Torchao Docs <https://huggingface.co/docs/transformers/main/en/quantization/torchao>`_.
+
+Inference with vLLM
```
Comment (jerryzh168) [inline docs/source/inference.rst]: for this section, can you replace with https://huggingface.co/pytorch/Qwen3-8B-int4wo-hqq#inference-with-vllm

it might be easier to do command line compared to code

File: docs/source/serving.rst [hunk 14]
```diff
@@ -9,4 +9,530 @@ serving step.
...
+        # Print the outputs.
+        print("\nGenerated Outputs:\n" + "-" * 60)
+        for output in outputs:
+            prompt = output.prompt
+            generated_text = output.outputs[0].text
+            print(f"Prompt:    {prompt!r}")
+            print(f"Output:    {generated_text!r}")
+            print("-" * 60)
+
+[Optional] Inference with Transformers
```
Comment (drisspg) [inline docs/source/serving.rst]: We should have an Inference w/ SGlang section

Comment (jainapurva) [reply docs/source/serving.rst]: I tested the integration of TorchAO and SGLang, came across a lot of issues in running the server. As discussed with @jerryzh168 offline, we can add this later, after more thorough testing and updates.

File: docs/source/serving.rst [hunk 15]
```diff
@@ -1,12 +1,468 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+
+    # Push to hub
+    USER_ID = "YOUR_USER_ID"
+    MODEL_NAME = model_id.split("/")[-1]
+    save_to = f"{USER_ID}/{MODEL_NAME}-float8dq"
+    quantized_model.push_to_hub(save_to, safe_serialization=False)
+    tokenizer.push_to_hub(save_to)
+
+    # Manual Testing
```
Comment (andrewor14) [inline docs/source/serving.rst]: I would split this into a separate code block and add some text in between, since everything below this line is technically not part of the user flow

File: docs/source/serving.rst [hunk 16]
```diff
@@ -1,12 +1,468 @@
...
+    print("Templated prompt:", templated_prompt)
+    inputs = tokenizer(
+        templated_prompt,
+        return_tensors="pt",
+    ).to("cuda")
+    generated_ids = quantized_model.generate(**inputs, max_new_tokens=128)
+    output_text = tokenizer.batch_decode(
+        generated_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
+    )
+    print("Response:", output_text[0][len(prompt):])
```
Comment (andrewor14) [inline docs/source/serving.rst]: Can you also add an example of what is printed here?

File: docs/source/serving.rst [hunk 17]
```diff
@@ -1,12 +1,468 @@
...
+    pip install accelerate
+
+.. code-block:: python
+
+    import torch
+    from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig
+
+    model_id = "microsoft/Phi-4-mini-instruct"
+
+    from torchao.quantization import Float8DynamicActivationFloat8WeightConfig, PerRow
```
Comment (andrewor14) [inline docs/source/serving.rst]: nit: move this to the top like the other imports?

File: docs/source/serving.rst [hunk 18]
```diff
@@ -1,12 +1,468 @@
...
+vLLM automatically leverages torchao's optimized kernels when serving quantized models, providing significant throughput improvements.
+
+First, install vLLM with torchao support:
+
+.. code-block:: bash
+
+    pip install vllm --pre --extra-index-url https://wheels.vllm.ai/nightly
+    pip install --pre torchao --index-url https://download.pytorch.org/whl/nightly/cu126
+
+.. code-block:: bash
```
Comment (andrewor14) [inline docs/source/serving.rst:68]: I think we need to add some text here. E.g. we need to explain we're serving with the quantized checkpoint we pushed to HF hub above. Also would be good to clarify that "float8dq" stands for "dynamic quant"

File: docs/source/serving.rst [hunk 19]
```diff
@@ -1,12 +1,468 @@
...
+    curl http://localhost:8000/v1/chat/completions -H "Content-Type: application/json" -d '{
+    "model": "appy1234/Phi-4-mini-instruct-float8dq",
+    "messages": [
+        {"role": "user", "content": "Give me a short introduction to large language models."}
+    ],
+    "temperature": 0.6,
+    "top_p": 0.95,
+    "top_k": 20,
+    "max_tokens": 32768
+    }'
```
Comment (andrewor14) [inline docs/source/serving.rst:83]: I think we should also quickly summarize that serving the float8 with vLLM is X times faster than serving the original high precision model, e.g. from the model card: https://huggingface.co/pytorch/Phi-4-mini-instruct-float8dq#quantization-recipe

Serve using vLLM with 36% VRAM reduction, 1.15x-1.2x speedup and little to no accuracy impact on H100.

File: docs/source/serving.rst [hunk 20]
```diff
@@ -1,12 +1,468 @@
...
+        "do_sample": False,
+    }
+
+    output = pipe(messages, **generation_args)
+    print(output[0]['generated_text'])
+
+Mobile Deployment with ExecuTorch
+--------------------------------
+
+ExecuTorch enables on-device inference using torchao's mobile-optimized quantization schemes. The 8da4w (8-bit dynamic activation, 4-bit weight) configuration is specifically designed for mobile deployment.
```
Comment (andrewor14) [inline docs/source/serving.rst]: I think we should link to part 2 (fine-tuning) here to say, for example, you can lower the QAT model there with the same quantization scheme, like you can do the below with or without QAT, so it's more like an optional step in an end-to-end flow. Right now this feels pretty separate

File: docs/source/serving.rst [hunk 21]
```diff
@@ -1,12 +1,468 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We want to quantize the embedding and lm_head differently. Since those layers are tied, we first need to untie the model:
+
+.. code-block:: python
+
+    from transformers import (
+    AutoModelForCausalLM,
+    AutoProcessor,
+    AutoTokenizer,
```
Comment (andrewor14) [inline docs/source/serving.rst]: nit: formatting is off, here and other code blocks below

File: docs/source/serving.rst [hunk 22]
```diff
@@ -1,12 +1,468 @@
...
+    )
+    import torch
+
+    model_id = "microsoft/Phi-4-mini-instruct"
+    untied_model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype="auto", device_map="auto")
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+
+    print(untied_model)
+    from transformers.modeling_utils import find_tied_parameters
```
Comment (andrewor14) [inline docs/source/serving.rst]: nit: move import to top

File: docs/source/serving.rst [hunk 23]
```diff
@@ -1,12 +1,468 @@
...
+
+Mobile Deployment with ExecuTorch
+--------------------------------
+
+ExecuTorch enables on-device inference using torchao's mobile-optimized quantization schemes. The 8da4w (8-bit dynamic activation, 4-bit weight) configuration is specifically designed for mobile deployment.
+
+Step 1: Untie Embedding Weights
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We want to quantize the embedding and lm_head differently. Since those layers are tied, we first need to untie the model:
```
Comment (andrewor14) [inline docs/source/serving.rst]: Is this step necessary actually? I don't think I had to do any of this for Llama models for example. Can you share the source for this?

File: docs/source/serving.rst [hunk 24]
```diff
@@ -1,12 +1,468 @@
...
+    from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig
+
+    # use "microsoft/Phi-4-mini-instruct" or "pytorch/Phi-4-mini-instruct-float8dq"
+    model_id = "pytorch/Phi-4-mini-instruct-float8dq"
+    quantized_model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", torch_dtype=torch.bfloat16)
+    tokenizer = AutoTokenizer.from_pretrained(model_id)
+
+    torch.cuda.reset_peak_memory_stats()
+
+    prompt = "Hey, are you conscious? Can you talk to me?"
```
Comment (andrewor14) [inline docs/source/serving.rst:324]: I see this code duplicated 3 times. Can we just have it appear in 1 place? Having it under "Evaluation" makes sense to me

File: docs/source/serving.rst [hunk 25]
```diff
@@ -1,12 +1,468 @@
...
+
+    # Evaluate baseline model
+    lm_eval --model hf --model_args pretrained=microsoft/Phi-4-mini-instruct --tasks hellaswag --device cuda:0 --batch_size 8
+
+    # Evaluate torchao-quantized model (float8dq)
+    lm_eval --model hf --model_args pretrained=pytorch/Phi-4-mini-instruct-float8dq --tasks hellaswag --device cuda:0 --batch_size 8
+
+Memory Benchmarking
+^^^^^^^^^^^^^^^^^
+
```
Comment (andrewor14) [inline docs/source/serving.rst:311]: Should add some text here. "For the Phi-4-mini-instruct model, serving with float8 dynamic quantization used X% less memory" or something

**Review by andrewor14 — APPROVED** [review]
Looks great, overall I feel we should add some more text in between code blocks so it feels more like a tutorial, and remove some duplicate code, which is distracting to readers

File: docs/source/serving.rst (hunk 14 above)
Comment (andrewor14) [reply docs/source/serving.rst]: Should we at least add an SGLang section and say `(Coming soon!)` or something? It's in the diagram at the top right now so people may search for it

File: docs/source/serving.rst (hunk 23 above)
Comment (jainapurva) [reply docs/source/serving.rst]: I'm using the same steps as here: https://huggingface.co/pytorch/Phi-4-mini-instruct-8da4w. In case of any updates, we should update both the model card and tutorial with same instructions
